feat(typescript)/dynamic template

iimez opened the pull request:
## Describe your changes

## Issue ticket number and link

## Checklist before requesting a review
- [ ] I have performed a self-review of my code.
- [ ] If it is a core feature, I have added thorough tests.
- [ ] I have added thorough documentation for my code.
- [ ] I have tagged PR with relevant project labels. I acknowledge that a PR without labels may be dismissed.
- [ ] If this PR addresses a bug, I have provided both a screenshot/video of the original bug and the working solution.

## Demo
<!-- Screenshots or video of new or updated code changes !-->

### Steps to Reproduce
<!-- Steps to reproduce demo !-->

## Notes
- InferenceModel & EmbeddingModel wrappers to hold some state in JS
- simplified downloadModel a little
- retrieveModel now returns the model config from the downloaded json
- createPrompt -> formatChatPrompt and updated it to use promptTemplate & systemPrompt
- added py defaults as defaults to createCompletion
- removed ModelFile types
- removed the extra yarn install as the issues it fixed seems to be solved

some work still needed:
- kv cache reuse / proper cache session (and context reset if its not a chat session)
- better types for `loadModel`
- regarding `downloadModel`: prompt templates not getting used when `allowDownload=false` might be a gotcha and make models suddenly perform worse when used offline. we could cache models.json in the `modelsPath` maybe?

jacoobes commented on gpt4all-bindings/typescript/src/gpt4all.d.ts:
```diff
@@ -8,41 +8,70 @@ type ModelType = "gptj" | "llama" | "mpt" | "replit";
 /**
  * Full list of models available
```
Deprecate instead of removing

jacoobes commented on gpt4all-bindings/typescript/src/gpt4all.d.ts:
```diff
@@ -8,41 +8,70 @@ type ModelType = "gptj" | "llama" | "mpt" | "replit";
+    
+    embed(
+        text: string,
+    ): Promise<Float32Array>;
```
fix as well, it shouldn't be async

iimez commented:
some update notes:
- reverted the `embed()` changes (removed promises)
- used overloads to type loadModel, added ModelFile type back in and added @deprecation notice
- createCompletion now takes a `systemPromptTemplate` option that allows for support of system messages. this arg is optional because i couldnt find a good default value yet. this means we cannot include system messages without the user explicitly specifying `systemPromptTemplate`. code is currently issueing a `console.warn` in this case. maybe we cant find a suitable default value.
- createCompletion now takes `promptHeader` and `promptFooter` options, replacements for `hasDefaultHeader` and `hasDefaultFooter` because these dont exist anymore in that sense. code is issueing a warning here as well.

getting good results from snoozy, mpt chat and falcon.

[batman-chat.md](https://github.com/nomic-ai/gpt4all/files/12206249/batman-chat.md)

iimez commented:
update notes:
- added `normalizePromptContext` method that allows to pass promptContext to InferenceModel.generate using either camelcase or snakecase. also added a test for it.
- fixed some libs amiss when doing `yarn build:backend` on linux

iimez commented:
update notes:
- added support for specifying a local models.json file in `loadModel`. if `modelConfigFile` is specified it will be used instead of the remote one. this allows for easier use of the bindings offline. ty @cosmic-snow 
- outputting a warning now if theres no valid `modelConfigFile` and `allowDownload: false`
- included some excellent prompt context docs by @Amogus8P

jacoobes commented on gpt4all-bindings/typescript/src/util.js:
```diff
@@ -1,10 +1,11 @@
+
 async function listModels() {
     const res = await fetch("https://gpt4all.io/models/models.json");
     const modelList = await res.json();
```
we can add some error handling here. 
```ts
if(res.ok) {
  return res.json()
} else {
  throw Error("...");
}
```

jacoobes commented on gpt4all-bindings/typescript/src/util.js:
```diff
@@ -32,10 +33,26 @@ function readChunks(reader) {
+    const normalizedPromptContext = {};
+
+    for (const key in promptContext) {
+        if (Object.prototype.hasOwnProperty.call(promptContext, key)) {
```
line 40: promptContext.hasOwnProperty(key)

jacoobes replied:
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/hasOwnProperty

jacoobes commented on gpt4all-bindings/typescript/src/util.js:
```diff
@@ -69,65 +87,66 @@ function downloadModel(modelName, options = {}) {
+            .then((res) => {
+                if (!res.ok) {
+                    reject(
+                        Error(
```
im starting to hate this absurd thing that prettier does haha we might need to switch or alter rules

jacoobes commented on gpt4all-bindings/typescript/test/gpt4all.test.js:
```diff
@@ -1,79 +1,122 @@
+        ...actualModule,
+        downloadModel: jest.fn(() => ({
+            cancel: jest.fn(),
+            promise: jest.fn(),
```
might need to alter the test to rrepresent promise not being a function anymore

iimez commented:
more changes:
- a deprecation warning gets output now if prompt context is defined with snake case
- added models.json url and the default prompt context values to config, and main exports
- use less defensive hasOwnProperty call
- moved the logic for picking remote json or local file into `listModels`. and added a check for `res.ok` in the fetch call there.
- some tests for `downloadModel` and `listModels`

jacoobes commented on gpt4all-bindings/typescript/src/util.js:
```diff
@@ -4,12 +4,35 @@ const { performance } = require("node:perf_hooks");
-    const res = await fetch("https://gpt4all.io/models/models.json");
-    const modelList = await res.json();
-    return modelList;
+    return [];
```
Should this be intended behavior? 🤔  I think we should throw if options nor url was provided

iimez replied:
yea not great. added a check for missing source option added a test checking for that behavior
